crates/hir/src/semantics.rs

```diff
@@ -155,6 +157,28 @@ impl<'db, DB: HirDatabase> Semantics<'db, DB> {
         self.imp.ancestors_at_offset_with_macros(node, offset)
     }
 
+    /// Iterates all `ModuleDef`s and `Impl` blocks of the given file.
+    pub fn visit_file_defs(&self, file_id: FileId, cb: &mut dyn FnMut(Either<ModuleDef, Impl>)) {
+        let module = match self.to_module_def(file_id) {
+            Some(it) => it,
+            None => return,
+        };
+        let mut defs: VecDeque<_> = module.declarations(self.db).into();
+        while let Some(def) = defs.pop_front() {
+            if let ModuleDef::Module(submodule) = def {
+                if let ModuleSource::Module(_) = submodule.definition_source(self.db).value {
+                    defs.extend(submodule.declarations(self.db));
+                    submodule
+                        .impl_defs(self.db)
+                        .into_iter()
+                        .for_each(|impl_| cb(Either::Right(impl_)));
+                }
+            }
+            cb(Either::Left(def));
+        }
+        module.impl_defs(self.db).into_iter().for_each(|impl_| cb(Either::Right(impl_)));
+    }
+
```

Comment:
Is semantics the best spot to put this in? Was really unsure where this belongs

Reply:
I think this belongs to `ide_db` (or maybe just `ide`). I don't think semancis should be really aware about the division of code into files, it is IDE specific concern.

Reply:
Ye that sounds like a better fit, moved it to `ide_db` now